server/src/main/java/org/apache/druid/catalog/model/table/InputFormats.java

```diff
@@ -0,0 +1,301 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Definition of the input formats which converts from property
+ * lists in table specs to subclasses of {@link InputFormat}.
+ */
+public class InputFormats
+{
+  public interface InputFormatDefn
+  {
+    String name();
+    String typeTag();
+    List<PropertyDefn<?>> properties();
+    void validate(ResolvedTable table);
+    InputFormat convert(ResolvedTable table);
+  }
+
+  public abstract static class BaseFormatDefn implements InputFormatDefn
+  {
+    private final String name;
+    private final String typeTag;
+    private final List<PropertyDefn<?>> properties;
+
+    public BaseFormatDefn(
+        final String name,
+        final String typeTag,
+        final List<PropertyDefn<?>> properties
+    )
+    {
+      this.name = name;
+      this.typeTag = typeTag;
+      this.properties = properties;
+    }
+
+    @Override
+    public String name()
+    {
+      return name;
+    }
+
+    @Override
+    public String typeTag()
+    {
+      return typeTag;
+    }
+
+    @Override
+    public List<PropertyDefn<?>> properties()
+    {
+      return properties;
+    }
+
+    @Override
+    public void validate(ResolvedTable table)
+    {
+      convert(table);
+    }
+  }
+
+  /**
+   * Definition of a flat text (CSV and delimited text) input format.
+   * <p>
+   * Note that not all the fields in
+   * {@link org.apache.druid.data.input.impl.FlatTextInputFormat
+   * FlatTextInputFormat} appear here:
+   * <ul>
+   * <li>{@code findColumnsFromHeader} - not yet supported in MSQ.</li>
+   * <li>{@code hasHeaderRow} - Always set to false since we don't bother to read
+   * it. {@code skipHeaderRows} is used to specify the number of header
+   * rows to skip.</li>
+   * </ul>
+   */
+  public abstract static class FlatTextFormatDefn extends BaseFormatDefn
+  {
+    public static final String LIST_DELIMITER_PROPERTY = "listDelimiter";
+    public static final String SKIP_ROWS_PROPERTY = "skipRows";
+
+    public FlatTextFormatDefn(
+        final String name,
+        final String typeTag,
+        final List<PropertyDefn<?>> properties
+    )
+    {
+      super(
+          name,
+          typeTag,
+          CatalogUtils.concatLists(
+              Arrays.asList(
+                  new StringPropertyDefn(LIST_DELIMITER_PROPERTY),
+                  new IntPropertyDefn(SKIP_ROWS_PROPERTY)
+              ),
+              properties
+          )
+      );
+    }
+
+    protected Map<String, Object> gatherFields(ResolvedTable table)
+    {
+      Map<String, Object> jsonMap = new HashMap<>();
+      jsonMap.put(InputFormat.TYPE_PROPERTY, CsvInputFormat.TYPE_KEY);
+      jsonMap.put("listDelimiter", table.property(LIST_DELIMITER_PROPERTY));
+       // hasHeaderRow is required, even though we don't infer headers.
+      jsonMap.put("hasHeaderRow", false);
+      jsonMap.put("findColumnsFromHeader", false);
+            // Column list is required. Infer from schema.
```

Comment:
Does this mean that the order of table columns must match the order of columns in the input? This seems like pretty big limitation.

Reply:
Didnt realize that MSQ does not support inferring columns from headers. When it does, maybe should revisit this.